File: analyze_freshness.py
#!/usr/bin/env python3
"""Analyze freshness result logs and print grouped statistics by file."""

import argparse
import re
import sys
from dataclasses import dataclass
from pathlib import Path
from typing import Dict, Iterable, List, Optional, Sequence


LINE_PATTERN = re.compile(
    r"^\[(?P<timestamp>[^\]]+)\] "
    r"\[(?P<database>[^.]+)\.(?P<table>[^\]]+)\] "
    r"freshness reached "
    r"elapsed=(?P<elapsed>\d+(?:\.\d+)?)s "
    r"baseline_row_count=(?P<baseline>\d+) "
    r"imported_rows=(?P<imported>\d+) "
    r"visible_rows=(?P<visible>\d+) "
    r"total_rows=(?P<total>\d+)$"
)


@dataclass
class FreshnessEntry:
    elapsed: float
    baseline_row_count: int
    imported_rows: int
    visible_rows: int
    total_rows: int


@dataclass
class InvalidLine:
    file_path: Path
    line_number: int
    line: str
    reason: str


@dataclass
class GroupStats:
    elapsed_sum: float = 0.0
    visible_sum: int = 0
    lines: int = 0

    def add(self, entry: FreshnessEntry) -> None:
        self.elapsed_sum += entry.elapsed
        self.visible_sum += entry.visible_rows
        self.lines += 1

    @property
    def avg_elapsed(self) -> float:
        return self.elapsed_sum / self.lines

    @property
    def avg_inserted(self) -> float:
        return self.visible_sum / self.lines


@dataclass
class FileStats:
    grouped: Dict[str, GroupStats]
    invalid_lines: List[InvalidLine]


def parse_args(argv: Sequence[str]) -> argparse.Namespace:
    parser = argparse.ArgumentParser(
        description="Analyze freshness_result_*.log files and print grouped statistics by file.",
    )
    parser.add_argument(
        "--log-dir",
        default="log",
        help="Directory that contains freshness result logs. Default: %(default)s",
    )
    parser.add_argument(
        "--pattern",
        default="freshness_result_*.log",
        help="Glob pattern used under --log-dir. Default: %(default)s",
    )
    parser.add_argument(
        "--show-invalid",
        action="store_true",
        help="Print invalid lines where total_rows != baseline_row_count + imported_rows.",
    )
    if not argv:
        parser.print_help()
        raise SystemExit(0)
    return parser.parse_args(argv)


def iter_log_files(log_dir: Path, pattern: str) -> List[Path]:
    return sorted(path for path in log_dir.glob(pattern) if path.is_file())


def parse_entry(line: str) -> Optional[FreshnessEntry]:
    match = LINE_PATTERN.match(line)
    if match is None:
        return None
    return FreshnessEntry(
        elapsed=float(match.group("elapsed")),
        baseline_row_count=int(match.group("baseline")),
        imported_rows=int(match.group("imported")),
        visible_rows=int(match.group("visible")),
        total_rows=int(match.group("total")),
    )


def group_name(baseline_row_count: int) -> str:
    bucket_floor = (baseline_row_count // 10000) * 10000
    return f"group-{bucket_floor}"


def analyze_file(file_path: Path) -> FileStats:
    grouped: Dict[str, GroupStats] = {}
    invalid_lines: List[InvalidLine] = []

    with file_path.open("r", encoding="utf-8") as handle:
        for line_number, raw_line in enumerate(handle, start=1):
            line = raw_line.rstrip("\n")
            entry = parse_entry(line)
            if entry is None:
                continue
            if entry.total_rows != entry.baseline_row_count + entry.imported_rows:
                invalid_lines.append(
                    InvalidLine(
                        file_path=file_path,
                        line_number=line_number,
                        line=line,
                        reason=(
                            "total_rows != baseline_row_count + imported_rows "
                            f"({entry.total_rows} != "
                            f"{entry.baseline_row_count} + {entry.imported_rows})"
                        ),
                    )
                )
                continue
            current_group = group_name(entry.baseline_row_count)
            grouped.setdefault(current_group, GroupStats()).add(entry)

    return FileStats(
        grouped=grouped,
        invalid_lines=invalid_lines,
    )


def analyze_logs(file_paths: Iterable[Path]) -> Dict[Path, FileStats]:
    results: Dict[Path, FileStats] = {}
    for file_path in file_paths:
        results[file_path] = analyze_file(file_path)
    return results


def merge_grouped_stats(results: Dict[Path, FileStats]) -> Dict[str, GroupStats]:
    merged: Dict[str, GroupStats] = {}
    for file_stats in results.values():
        for name, stats in file_stats.grouped.items():
            target = merged.setdefault(name, GroupStats())
            target.elapsed_sum += stats.elapsed_sum
            target.visible_sum += stats.visible_sum
            target.lines += stats.lines
    return merged


def sorted_group_names(grouped: Dict[str, GroupStats]) -> List[str]:
    return sorted(grouped, key=lambda item: int(item.split("-", 1)[1]))


def print_grouped_stats(title: str, grouped: Dict[str, GroupStats]) -> None:
    print(title)
    print("起始行数量级, 平均插入行数, 统计条目数, 平均耗时(s)")
    for name in sorted_group_names(grouped):
        group_stats = grouped[name]
        print(
            f"{name}, {group_stats.avg_inserted:.2f}, "
            f"{group_stats.lines}, {group_stats.avg_elapsed:.2f}"
        )
    print()


def main() -> int:
    args = parse_args(sys.argv[1:])
    log_dir = Path(args.log_dir)
    file_paths = iter_log_files(log_dir, args.pattern)

    if not file_paths:
        print(f"no log files found in {log_dir} matching {args.pattern}")
        return 1

    results = analyze_logs(file_paths)

    for file_path in file_paths:
        stats_by_file = results[file_path]
        print_grouped_stats(file_path.name, stats_by_file.grouped)
        if args.show_invalid and stats_by_file.invalid_lines:
            print("invalid_details")
            for item in stats_by_file.invalid_lines:
                print(f"{item.file_path}:{item.line_number}: {item.reason}")
                print(item.line)
        print()

    print_grouped_stats("Summarize", merge_grouped_stats(results))

    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: src/bin/analyze_freshness.rs
use clap::Parser;
use std::collections::BTreeMap;
use std::path::{Path, PathBuf};
use tici_test_rust::common::resolve_project_path;

#[derive(Parser, Debug)]
#[command(about = "Analyze freshness_result_*.log files and print grouped statistics by file.")]
struct Cli {
    #[arg(long, default_value = "log")]
    log_dir: String,
    #[arg(long, default_value = "freshness_result_*.log")]
    pattern: String,
    #[arg(long, default_value_t = false)]
    show_invalid: bool,
}

#[derive(Clone)]
struct FreshnessEntry {
    elapsed: f64,
    baseline_row_count: u64,
    imported_rows: u64,
    visible_rows: u64,
    total_rows: u64,
}

#[derive(Clone)]
struct InvalidLine {
    file_path: PathBuf,
    line_number: usize,
    line: String,
    reason: String,
}

#[derive(Default, Clone)]
struct GroupStats {
    elapsed_sum: f64,
    visible_sum: u64,
    lines: u64,
}

impl GroupStats {
    fn add(&mut self, entry: &FreshnessEntry) {
        self.elapsed_sum += entry.elapsed;
        self.visible_sum += entry.visible_rows;
        self.lines += 1;
    }
}

#[derive(Default, Clone)]
struct FileStats {
    grouped: BTreeMap<String, GroupStats>,
    invalid_lines: Vec<InvalidLine>,
}

fn parse_entry(line: &str) -> Option<FreshnessEntry> {
    let line = line.trim();
    let (_, tail) = line.split_once("] [")?;
    let (_, tail) = tail.split_once("] freshness reached elapsed=")?;
    let (elapsed, tail) = tail.split_once("s baseline_row_count=")?;
    let (baseline, tail) = tail.split_once(" imported_rows=")?;
    let (imported, tail) = tail.split_once(" visible_rows=")?;
    let (visible, total) = tail.split_once(" total_rows=")?;
    Some(FreshnessEntry {
        elapsed: elapsed.trim().parse().ok()?,
        baseline_row_count: baseline.trim().parse().ok()?,
        imported_rows: imported.trim().parse().ok()?,
        visible_rows: visible.trim().parse().ok()?,
        total_rows: total.trim().parse().ok()?,
    })
}

fn group_name(baseline_row_count: u64) -> String {
    format!("group-{}", (baseline_row_count / 10000) * 10000)
}

fn analyze_file(file_path: &Path) -> Result<FileStats, String> {
    let content = std::fs::read_to_string(file_path).map_err(|err| err.to_string())?;
    let mut stats = FileStats::default();
    for (idx, raw_line) in content.lines().enumerate() {
        let Some(entry) = parse_entry(raw_line) else {
            continue;
        };
        if entry.total_rows != entry.baseline_row_count + entry.imported_rows {
            stats.invalid_lines.push(InvalidLine {
                file_path: file_path.to_path_buf(),
                line_number: idx + 1,
                line: raw_line.to_string(),
                reason: format!(
                    "total_rows != baseline_row_count + imported_rows ({} != {} + {})",
                    entry.total_rows, entry.baseline_row_count, entry.imported_rows
                ),
            });
            continue;
        }
        stats
            .grouped
            .entry(group_name(entry.baseline_row_count))
            .or_default()
            .add(&entry);
    }
    Ok(stats)
}

fn print_grouped_stats(title: &str, grouped: &BTreeMap<String, GroupStats>) {
    println!("## {title}");
    println!("| 起始行数量级 | 平均插入行数 | 统计条目数 | 平均耗时(s) |");
    println!("| --- | ---: | ---: | ---: |");
    for (name, stats) in grouped {
        let avg_inserted = stats.visible_sum as f64 / stats.lines as f64;
        let avg_elapsed = stats.elapsed_sum / stats.lines as f64;
        println!(
            "| {name} | {avg_inserted:.2} | {} | {avg_elapsed:.2} |",
            stats.lines
        );
    }
    println!();
}

fn print_invalid_lines(invalid_lines: &[InvalidLine]) {
    println!("### invalid_details");
    println!("| 文件 | 行号 | 原因 | 原始内容 |");
    println!("| --- | ---: | --- | --- |");
    for item in invalid_lines {
        println!(
            "| {} | {} | {} | {} |",
            item.file_path.display(),
            item.line_number,
            item.reason.replace('|', r"\|"),
            item.line.replace('|', r"\|")
        );
    }
    println!();
}

fn main() {
    let cli = Cli::parse();
    let log_dir = resolve_project_path(&cli.log_dir);
    let pattern =
        glob::Pattern::new(&cli.pattern).unwrap_or_else(|_| glob::Pattern::new("*").unwrap());
    let mut file_paths = match std::fs::read_dir(&log_dir) {
        Ok(entries) => entries
            .filter_map(|entry| entry.ok())
            .map(|entry| entry.path())
            .filter(|path| path.is_file())
            .filter(|path| {
                path.file_name()
                    .and_then(|name| name.to_str())
                    .map(|name| pattern.matches(name))
                    .unwrap_or(false)
            })
            .collect::<Vec<_>>(),
        Err(_) => Vec::new(),
    };
    file_paths.sort();
    if file_paths.is_empty() {
        println!(
            "no log files found in {} matching {}",
            log_dir.display(),
            cli.pattern
        );
        std::process::exit(1);
    }

    let mut merged = BTreeMap::<String, GroupStats>::new();
    for file_path in &file_paths {
        let stats = match analyze_file(file_path) {
            Ok(value) => value,
            Err(err) => {
                eprintln!("{err}");
                std::process::exit(2);
            }
        };
        print_grouped_stats(
            file_path
                .file_name()
                .and_then(|item| item.to_str())
                .unwrap_or_default(),
            &stats.grouped,
        );
        if cli.show_invalid && !stats.invalid_lines.is_empty() {
            print_invalid_lines(&stats.invalid_lines);
        }
        for (name, group) in stats.grouped {
            let target = merged.entry(name).or_default();
            target.elapsed_sum += group.elapsed_sum;
            target.visible_sum += group.visible_sum;
            target.lines += group.lines;
        }
    }
    print_grouped_stats("Summarize", &merged);
}
